### ISSUE TITLE
Errorness Transformation of Objects

### ISSUE THREAD

**Issue by FYRichie** [facebookresearch/sam-3d-objects#56]
Hi, thank you for this amazing work!

I have a question about the object transformation from `canonical` to `camera` space. I have an indoor scene image (from HyperSim) which looks like this:
![Image](https://github.com/user-attachments/assets/b523fdaf-dc8f-4189-87c9-1880e5d0dccf)

The results I got from the online demo is pretty good (shown in the image below) 
<img width="2356" height="1094" alt="Image" src="https://github.com/user-attachments/assets/6360a1e0-c1d3-41ee-8516-f5d2b04034d7" />

However, when I tested the released code and followed `demo_multi_object.ipynb`, the result (rendered GIF) looks very differently to the online demo.
<img width="510" height="511" alt="Image" src="https://github.com/user-attachments/assets/46aef7d3-b94e-4f58-b7ba-3d75158344c4" />

Also, when I transform the output glb mesh with the estimated scale, rotation, and translation, the estimation doesn't seem to be accurate (visualized in blender, figure below).
<img width="774" height="701" alt="Image" src="https://github.com/user-attachments/assets/be143b78-91fd-4d99-85d9-ef39d9aa5ba7" />

I think it's reasonable when the object is cropped (ex. the towel near the camera, the metal ring), but I don't understand why objects that are complete (ex. the brown box, duck, red bottle, bathtub) have weird spatial relationships. Is there anything that I'm doing wrong.

Again, thanks for your work, and I'll be grateful if you can provide any suggestions.

**Comment by JiahaoXia** [comment]
It would be great if the authors could clarify the coordinate system they are using when generating the 3D mesh.

**Comment by FYRichie** [comment]
Hi @JiahaoXia , do you think this is an issue caused by operating on the incorrect coordinate system? Because I didn't do any operation between the mesh generation and video (GIF) rendering. It follows the same code structure as the demo.

**Comment by alexsax** [comment]
Thanks for the detailed report!

In general, SAM 3D uses pytorch3d camera conventions: https://pytorch3d.org/docs/cameras.

The result in the online demo looks about like I would expect; but the local code appears a little bit worse. Are the same generations used in the rendered GIF and the Blender image? I ask because I think the GLB may be rotated or flipped 180 degrees from the 'object' coordinate space, which is expected.

However, the towel ring in the Blender image looks to be in a different location relative to the towel than it is in the rendered image from `demo_multi_object.inpynb`

**Comment by JiahaoXia** [comment]
> Hi [@JiahaoXia](https://github.com/JiahaoXia) , do you think this is an issue caused by operating on the incorrect coordinate system? Because I didn't do any operation between the mesh generation and video (GIF) rendering. It follows the same code structure as the demo.

I am experiencing a similar issue aligning the 3D mesh with the original image, and I believe it would be helpful if the authors of SAM3D could clarify the coordinate system they use.

**Comment by alexsax** [comment]
Hi @JiahaoXia,

The predictions are in Pytorch3d camera conventions:
> In general, SAM 3D uses pytorch3d camera conventions: https://pytorch3d.org/docs/cameras.

You can see how the transform is applied to take the predicted `gaussians` from canonical space to the camera space, via this `make_scene` [function](https://github.com/facebookresearch/sam-3d-objects/blob/cf066761706cd02b07e2fc6274570ec8cdafb683/notebook/inference.py#L263-L268) that's used in `demo_multi_object.ipynb`

Do you find that the mesh you are using is not aligned with the `gaussians` returned during model inference?

**Comment by FYRichie** [comment]
Hi @alexsax , thanks for your reply! The code I used to render the GIF and to get the mesh is like this:
```python
outputs = []
for mask in all_masks:
    output = inference(img, mask, 42, pointmap)

    # for rendering GIF
    outputs.append(output)

    # for mesh demo in blender
    mesh = output["glb"]
    scale = output["scale"].cpu().squeeze().numpy()
    rotation = np.zeros((4, 4))
    rotation[:3, :3] = quaternion_to_matrix(output["rotation"]).cpu().squeeze().numpy()
    rotation[3, 3] = 1
    translation = output["translation"].cpu().squeeze().numpy()
    mesh.apply_scale(scale).apply_transform(rotation).apply_translation(translation)
    mesh.export(save_path)

# render GIF
scene_gs = make_scene(*outputs)
scene_gs = ready_gaussian_for_video_rendering(scene_gs)
video = render_video(scene_gs, r=1, fov=60, resolution=512)["color"]
# save GIF
```
I follow the steps in `compose_transform` in `sam3d_objects/data/dataset/tdfy/transform_3d.py` to apply the transformations. This function is used in the `make_scene` function (in `SceneVisualizer.object_pointcloud`). There might be some other errors in my usage that I haven't found (sorry that I haven't go through your code yet).

My question is: Why is the GLB may be rotated or flipped 180 degrees from the 'object' coordinate space expected?

**Comment by codepk37** [comment]
@FYRichie Hi, I'm facing same issue, are you able to resolve it ?

**Comment by Constannnnnt** [comment]
Not sure if I understand this correct, but In the [to_glb](https://github.com/facebookresearch/sam-3d-objects/blob/cf066761706cd02b07e2fc6274570ec8cdafb683/sam3d_objects/model/backbone/tdfy_dit/utils/postprocessing_utils.py#L666), a GLB which is already Z-up→Y-up rotated, which means
```
obj["glb"]: mesh vertices are in local object space, already Z-up→Y-up rotated.
obj["translation"], obj["rotation"], obj["scale"]: layout in canonical/camera space before that Z-up→Y-up rotation.
```
So [`make_scene`](https://github.com/facebookresearch/sam-3d-objects/blob/cf066761706cd02b07e2fc6274570ec8cdafb683/notebook/inference.py#L255) and `to_glb` are in different frames:
```
Gaussians and layout → canonical / camera frame.
GLB vertices → canonical object frame, but with a fixed Y-up rotation baked in.
```

My solution is to transform the glb files back to z-up, apply the transformations from the gs outputs and then transform back to y-up.

**Comment by alexsax** [comment]
@Constannnnnt yes you have the right idea here, that the GLB mesh and Gaussian outputs use different coordinate frames. these are the coordinate frames of the different outputs:
  - **GLB mesh:** Exported in Y-up coordinates (standard for glTF/GLB format). The Z-up → Y-up rotation is applied in [`postprocessing_utils.py`](https://github.com/fairinternal/lidra/blob/01b65b3945e6931dd8d2ab5b7b08d93f8995950b/lidra/model/backbone/trellis/utils/postprocessing_utils.py#L665)
  - **Gaussian splats:** Stay in Z-up coordinates (no rotation applied in save_ply())
  - **Transformation parameters (rotation, translation, scale):** Computed in the original Z-up/PyTorch3D camera frame

So @FYRichie -- the upshot here is that if you apply the transformation parameters directly to GLB mesh vertices in Blender the mesh vertices have already been rotated to Y-up but the transforms haven't.
Here is an (untested) manual workaround to that might work in Blender
```python
# Manual workaround to convert the GLB mesh coordinate (Y-up → Z-up):
import numpy as np

R_zup_to_yup = np.array([[1, 0, 0], [0, 0, -1], [0, 1, 0]])
R_yup_to_zup = R_zup_to_yup.T

# 1. Convert GLB vertices back to Z-up
vertices_zup = glb_vertices @ R_yup_to_zup
# 2. Apply your transformation in Z-up frame
# 3. Convert back to Y-up
vertices_final = transformed_vertices @ R_zup_to_yup
```

**Comment by alexsax** [comment]
Here is a standalone snippet that you might find useful for combining `glb` meshes into a Y-up scene. The function `make_scene_untextured_mesh` is analogous to [`make_scene`](https://github.com/facebookresearch/sam-3d-objects/blob/cf066761706cd02b07e2fc6274570ec8cdafb683/notebook/inference.py#L255) -- except that the new function returns the scene in Y-up.

```python
  import numpy as np
  import torch
  from copy import deepcopy
  from pytorch3d.transforms import quaternion_to_matrix
  from sam3d_objects.data.dataset.tdfy.transforms_3d import compose_transform

  _R_ZUP_TO_YUP = np.array([[1, 0, 0], [0, 0, -1], [0, 1, 0]], dtype=np.float32)
  _R_YUP_TO_ZUP = _R_ZUP_TO_YUP.T


  def make_scene_untextured_mesh(*outputs, in_place=False):
      import trimesh

      if not in_place:
          outputs = [deepcopy(output) for output in outputs]

      all_meshes = []
      for output in outputs:
          mesh = output["glb"]
          if mesh is None:
              continue

          # GLB is Y-up, transforms are Z-up; convert, apply, convert back
          vertices = mesh.vertices.astype(np.float32) @ _R_YUP_TO_ZUP
          vertices_tensor = torch.from_numpy(vertices).float().to(output["rotation"].device)
          R_l2c = quaternion_to_matrix(output["rotation"])
          l2c_transform = compose_transform(
              scale=output["scale"],
              rotation=R_l2c,
              translation=output["translation"],
          )
          vertices = l2c_transform.transform_points(vertices_tensor.unsqueeze(0))
          mesh.vertices = vertices.squeeze(0).cpu().numpy() @ _R_ZUP_TO_YUP
          all_meshes.append(mesh)

      if not all_meshes:
          return None

      if len(all_meshes) == 1:
          return all_meshes[0]

      return trimesh.util.concatenate(all_meshes)
```

```
Usage:
outputs = [inference(image, mask, seed=42) for mask in masks]
scene_mesh = make_scene_untextured_mesh(*outputs)
scene_mesh.export("scene.glb")
```

And here is the visualization in meshlab
<img width="767" height="699" alt="Image" src="https://github.com/user-attachments/assets/57ffcec1-d9c2-4692-ae55-246c5fd09cc7" />

**Comment by Constannnnnt** [comment]
Thank you @alexsax! I have another question: since the gaussian outputs use a coorindate from the camera space, where if the camera is tilted, the whole scene and objects might be a bit tilted (even though their poses and positions look almost correctly from the camera view). Then, how do we recover a better pose? Take the image you provided for example, the whole scene is a bit tilted and the mattress and the sofa might not be on the same plane and the floating frame might not have a local up close to world up. In this case, do depth and camera parameters help (e.g. a depth-anything 3 model)? I haven't tried this (I used a naive method: use PCA to get principal axes per object and then readjust its orientation), but would love to have some insights from you? Thank you very much!

**Comment by alexsax** [comment]
I really like P3D's figure for different coordinate frames to anchor the discussion.
<img width="2000" height="800" alt="Image" src="https://github.com/user-attachments/assets/f3ab425d-3e07-42de-ba9f-12fa9a7d0695" />

So in the pic above I'm interpreting this as as "the model returns poses in the camera space, but how do we transform this to world space" -- is that right? :)

You could try estimating the ground plane, and using that to determine the direction of gravity (i.e. which way is up), and rotating that to align with your preferred axis. We have a simple [`estimate_plane`](https://github.com/facebookresearch/sam-3d-objects/blob/afdf6a31522d038c44c68a0bb57aa68827380797/sam3d_objects/pipeline/inference_pipeline_pointmap.py#L451) function in the source code as a starting point.

Even after global rotaitons, the different objects might still be slightly misaligned -- that's more due to the fact that the model makes object predictions one-at-a-time and there can still be errors. My guess is that sort of physics/geometry post-optimization could help here (e.g. [DSO](https://ruiningli.com/dso)). If you end up using a physics-based optimization I'd be very curious about the results.

**Comment by Constannnnnt** [comment]
Thank you @alexsax, much appreicated! I will let you know if I have some results! Actually I have tried to implement some methods based on SDF, like [CAST](https://arxiv.org/abs/2505.02836) and [Scenethesis](https://arxiv.org/abs/2505.02836), but my results were before sam3d-objects and I focused on the scene language at that time, which was not ideal if the scene was complicated and occluded with many objects. I will try to integrate them if possible and update it here once I have them. Thanks again for your help!

**Comment by aman190202** [comment]
Hi @Constannnnnt , thank you for your comment on the `PCA` approach ; Can you emphasize a little more on how do you use it to position the objects?

**Comment by Constannnnnt** [comment]
Hi @aman190202, this `PCA approach` does not position objects but alters their poses, and sometimes this approach fails. 

This approach simply just makes meshes upright to handle the drift caused by the camera angles and assumes that we have already found a surface to land these objects; in my case, it is a floor. Ideally, it can iteratively land objects on different surfaces.  

```python
def make_mesh_upright(mesh: trimesh.Trimesh) -> None:
    """
    In-place: after global floor alignment, make a single object upright.

    Steps:
      1. PCA on its vertices to get principal axes.
      2. Choose the axis most aligned with WORLD_UP as 'current up'.
      3. Compute minimal rotation that maps that axis to WORLD_UP.
      4. Rotate around the object's base center.
      5. Re-snap bottom to y = 0.
    """
    v = mesh.vertices.astype(np.float32)

    # PCA
    mean = v.mean(axis=0)
    X = v - mean
    cov = X.T @ X / X.shape[0]
    eigvals, eigvecs = np.linalg.eigh(cov)   # columns are eigenvectors

    axes = [eigvecs[:, i] for i in range(3)]
    # pick axis with largest alignment to WORLD_UP
    dots = [abs(np.dot(a, WORLD_UP)) for a in axes]
    up_axis = axes[int(np.argmax(dots))]
    dot_up = float(np.dot(up_axis, WORLD_UP))
    # If nothing is even moderately aligned with up, avoid aggressive flips (e.g., trees with sideways branches)
    if abs(dot_up) < 0.5:
        y = v[:, 1]
        v[:, 1] -= float(y.min())
        mesh.vertices = v
        return
    if np.dot(up_axis, WORLD_UP) < 0:
        up_axis = -up_axis

    # rotation that maps up_axis -> WORLD_UP
    R_fix = rotation_from_normal_to_up(up_axis)

    # rotate around base center
    y = v[:, 1]
    y_min = float(y.min())
    base_mask = np.isclose(y, y_min, atol=0.02 * max(1.0, y.max() - y_min))
    if base_mask.sum() < 10:
        # fallback: use overall centroid projected to floor
        base_center = v.mean(axis=0)
        base_center[1] = y_min
    else:
        base_center = v[base_mask].mean(axis=0)

    v_shift = v - base_center[None, :]
    v_rot = v_shift @ R_fix.T
    v_new = v_rot + base_center[None, :]

    # re-snap to floor
    min_y_new = v_new[:, 1].min()
    v_new[:, 1] -= min_y_new

    mesh.vertices = v_new
```

**Comment by aman190202** [comment]
Hi Chen,

Thank you for your response and explanation. I am currently using
MapAnything to get a pointmap for the scene, which happens to be pretty
good but the objects are intersecting each other because of mask
inaccuracies I believe. Any ideas on how to resolve this?

**Comment by Constannnnnt** [comment]
@aman190202 Maybe this helps? https://github.com/facebookresearch/sam-3d-objects/pull/79

In case of mask inaccuracies, consider other models, maybe sam3 or sam-hq to obtain better masks?

**Comment by salihculcu** [comment]
Hello.
I was actively using Sam3D's 3D object creation playground.
However, there is currently a problem with the playground and I cannot create objects.

Is this temporary?
If not, I would like to rent a remote server and run it as open source.

What hardware is sufficient?
And are there any codes in this repository that allow me to perform segmentation like in Sam2, similar to the playground?

Thank you.
